rfc/system/0000-flyteadmin-rbac.md

```diff
@@ -0,0 +1,203 @@
+```yaml
+authorization:
+  policies: # policies is a list instead of a map since viper map keys are case insensitive
+    - name: admin
+      rules: # list of rules
+        - name: "allow all" # name / description of the rule
+          methodPattern: ".*" # regex pattern to match on gRPC method
+    - name: read-only
+      rules:
+        - name: "read everything"
+          methodPattern: "Get.*|List.*"
+    - name: mapping-team
+      rules:
+        - name: "r/w for the mapping project in dev only"
+          methodPattern: ".*"
+          project: mapping # access to only the mapping project
+          domain: development # access to only the development domain within the mapping project
+    - name: ci
+      rules:
+        - name: "r/w for every project in production"
+          methodPattern: ".*"
+          domain: production # you can wildcard project and declare domain level access
+    - name: 0oahjhk34aUxGnWcZ0h7 # the names can even include things like okta app IDs
+      rules:
+        - name: "flyte propeller"
+          methodPattern: ".*"
+```
+
+#### Authorization Interceptor 
+
+The authorization interceptor logic will behave like so: 
+1. The interceptor checks to see if the target RPC method is configured to be bypassed. Call the next handler if bypassed. 
+2. The interceptor consults the role resolver to obtain a list of role names for the user. 
+3. The interceptor iterates over the list of role names and searches for any authorization policies tied to the role names. If no matching authorization policies are found return Permission Denied.
+4. The interceptor iterates over the matching authorization policies to see if they permit the target RPC. Each rule that matches is added to a list. If no matching rules are found return Permission Denied. 
+5. The interceptor extracts the project and domain scope out of each of the permitting rules and sets authorized resource scope on the authorization context. 
+6. The interceptor calls the next handler.
+
+Below is a high level overview of what the authorization context would look like. 
+
+```go
+type ResourceScope struct {
+    Project string
+    Domain string
+}
+
+type AuthorizationContext struct {
+    resourceScopes []ResourceScope
+}
+
+// TargetResourceScope represents the scope of an individual resource. Sometimes only project level or domain level scope
+// is applicable to a resource. In such cases, a user's resource scope may need to be truncated to matcht he depth 
+// of the target resource scope.
+type TargetResourceScope = int
+
+const (
+    ProjectTargetResourceScope = 0
+    DomainTargetResourceScope = 1
+)
+```
+
+#### Authorization Utils + DB Layer
+
+The final piece of the puzzle is what performs resource level authorization and filtering. Historically, I have found 
+that the best (albeit challenging) way to do this is at the database layer for a few reasons: 
```

Comment:
This proposal makes sense, but there's some notable instances where we could have unauthorized side effects

specifically, for create execution, a user who has read access may be able to read the respective launch plan, we then create the CRD and only then insert the model in the DB (which gets rejected) here: https://github.com/flyteorg/flyte/blob/5f695895ed6a7fa45d980023ab874591184b0fc1/flyteadmin/pkg/manager/impl/execution_manager.go#L1027-L1062

so in this scenario, we create a rogue Flyte CRD that never gets committed to the db (operation fails with permission denied) that propeller will attempt to process and send event updates for, leading to potential system weirdness

Reply:
That is a good point, I wasn't aware of that interaction. 

I think there are a couple ways to work around this. 
1. Bring authorization for creation into the application layer (in addition to the db layer I suppose). 
2. Rework the code such that the execution is written to the database (maybe partially) prior to creating the CRD. And then writing the rest to the DB after the CRD is created if necessary. 

I am being a bit naive with the complexity to approach number 2 but I do think this is a general problem with distributed systems. For example, the way the code is written as is makes the system susceptible to ghost CRDs since the system could crash before writing to the DB. I also understand this issue is probably much more likely to happen with authorization errors though :)

The architectural issue I see with this type of behavior boils down to how business logic does what is effectively a distributed transaction across other services (submitting CRDs in this case) and writing to the database. The most robust way to attempt this sort of behavior that I've seen is to always write to the DB first (this is the source of truth) and then opportunistically do the rest of the distributed transaction. For cases where the server crashes after writing to the DB you'll basically need some anti entropy background task to handle edge cases.

Reply:
For sure, it's definitely a problem now even without authz checks failing. I recall we waffled on this approach initially because we didn't want to commit the entry to the db until after the CRD was created (since the workflowengine packages gracefully handles already exists on retries, but falsely implying an execution was created in the reversed case was more confusing)

if we do bring authorization to the application layer (even for these exceptional cases) I'm a bit concerned we're now depending on code review to enforce authorization checks in both the db and application levels